synapse/handlers/pagination.py

```diff
@@ -72,13 +73,123 @@ def __init__(self, hs):
         self.auth = hs.get_auth()
         self.store = hs.get_datastore()
         self.clock = hs.get_clock()
+        self.config = hs.config
 
         self.pagination_lock = ReadWriteLock()
         self._purges_in_progress_by_room = set()
         # map from purge id to PurgeStatus
         self._purges_by_id = {}
         self._event_serializer = hs.get_event_client_serializer()
 
+        if self.config.retention_enabled:
+            # Run the purge jobs described in the configuration file.
+            for job in self.config.retention_purge_jobs:
+                self.clock.looping_call(
+                    run_as_background_process,
+                    job["interval"],
+                    "purge_history_for_rooms_in_range",
+                    self.purge_history_for_rooms_in_range,
+                    job["shortest_max_lifetime"], job["longest_max_lifetime"],
+                )
+
+    @defer.inlineCallbacks
+    def purge_history_for_rooms_in_range(self, min_ts, max_ts):
+        """Purge outdated events from rooms within the given retention range.
+
+        The lower limit of this range is not inclusive, its upper limit is inclusive.
+
+        If a default retention policy is defined in the server's configuration and its
+        'max_lifetime' is within this range, also targets rooms which don't have a
+        retention policy.
+
+        Args:
+            min_ts (int): Timestamp in milliseconds that define the lower limit of the
+                range to handle.
+            max_ts (int): Timestamp in milliseconds that define the higher limit of the
+                range to handle.
+        """
+        # We want the storage layer to to include NULLs in its return value only if
+        # a default retention policy is defined in the server's configuration and that
+        # policy's 'max_lifetime' is either lower (or equal) than max_ts or higher than
+        # min_ts (or both).
+        if self.config.retention_default_max_lifetime is not None:
+            include_null = True
+
+            if min_ts is not None:
```

Comment:
I think this would be clearer:

```python
include_null = True

if min_ts is not None and min_ts >= self.config.retention_default_max_lifetime:
    # the default max_lifetime is lower than (or equal to) min_ts
    include_null = False

if max_ts is not None and max_ts < self.config.retention_default_max_lifetime:
    # the default max_lifetime is higher than max_ts
    include_null = False
```